### PR TITLE
TEP-0118: Update TaskRun Validation for Matrix Include Params

### PR DESCRIPTION
**Pull request by EmmaMunley**
# Changes

This PR updates TaskRun validation to also validate Matrix Include Params. With this change, we validate that all parameters declared in the TaskSpec are present in the TaskRun. It also validates that all parameters have values, parameter types match the specified type and object params have all the keys required.

Also added testing to bring test coverage to 100%.

Note: Matrix Include is still in preview mode.

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [N/A] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [N/A] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [N/A] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by EmmaMunley** [comment]
/Assign @jerop @pritidesai 
/Hold until #6407 merges
cc: @lbernick 

This is branched off of #6407 to avoid merge conflicts. 
Please only review these 2 files (the rest are part of  #6407)
`validate_taskrun.go`
`validate_taskrun_test.go`

**Comment by EmmaMunley** [comment]
Also not sure why the code coverage went down. I've compared the code coverage with the last PR to have file changes: #6150 and it looks like it has the exact same [code coverage](https://ffac19ab6c1d7f69330d45f61be13860ef12baf2383e8194558e81c-apidata.googleusercontent.com/download/storage/v1/b/tekton-prow/o/pr-logs%2Fpull%2Ftektoncd_pipeline%2F6150%2Fpull-tekton-pipeline-go-coverage-df%2F1633163535615791104%2Fartifacts%2Fline-cov.html?jk=AahUMlu239PxvELzbJzC_TAkhsKW2pYcTps9CQwkYDjBpJeayNOXZylnL7o4dky45nrD9dbKaDS1ElFV2wmYEnHbtehgs2XoFW6XK1FIWYpQQbIJIkyt6BiCnGz-WzDFXmViGW_HFkNptQcThofMiRThUIwKGhEcix3yCwaJtmt7Vln9QQqh2kNN8DTVXbXB6k14RkTGlagwRVKIBDV_-0Qaq4HErz-Det23TPN_sTgUFH76dZDBkjOsoe398iYmwsad4e2JLkqYMjOdbfvi6koY3troKg9Gj8gCJLu5M9PWuXyAOreMk_G7KqGk6of2DXMJ9nD9y5kUHxvczWIr4LGYoni5V1sge0BgN0GsctzRE1wItKXVUjb_REIOf54wEk5UUxNjy1H4055XsF6fXFxh5usS5HHy9SNxYOA6kgL6ZcYISQ64_mD8RH6fXqPp3gBtZKM1GE8ZpYoWeGE8l7kJXcVxpAa2--o47HUwHiY5_Ng1zyFtGMl-Tf7sHUDpY0qjWhR3PqJX13zaI06aCBPczw3aqUjM7Iic6bYVomBhpETJeUaPNds7miEMnpdR58Lza487DKdXkuzJKaIWNSqXMS3AR-i2nDrJOL8F4Ds_LhyOjDmufwfdMF096m0ANWjcM3DSgHkVDuFzzkqh8ZI4W9Jnp0SSOgiym9KAnH6LAcZsrPqyUs_2_W_rJPvLvas6r6JQ6dlLZG3Y4Oi3uSINQu6K-xHaQbxqcAPUNrH4BVpJa5sVON9xwMhB8_W7RJHxEhDv4_UUdjPVMvaOTkSF-92hKPucCb8ZbY7mP-536t7KIUmOHFGwa1JsmI4VVKfSgSIRWHJcRHcqAtOMexhKhO_oP-kQYpp6n7jdfvPj7MrPiTWgRsvZMz0LgymKtIeG2XAo7cJbZQScM72UyHXrgJpkFZbMe5yWTkt_gofz45pkScJKd5J2zI1R9Wljp70xK_5Rr1_3fXwt4dBwjeFKQAZqkzIillusJRqu2k10nC2vxa9JTZyU8JZAaOX9F0shjy8XpKpNV4GPK0sDaMP8fM8yULFcF5dSHmytclBc2YjxWUndFQNzp5zDXytYk9b3WUh8Q_jTkKFSOqlenfmRRP2QZKwhLMZYuFS51oqe1xWYb1OVcvkD--8KMmOat--N2OevKizQh2NU8bUyTMDlqlFpz1nLzpOW_nUKJ0RKej_iiF_ULix856p1aPlS0BhvXC9W6qDAGP1LRwR3Gwfkjsz0t7xdLI2BTkUMgGOCKVI5TuhP5yhNlbpFjLkbq9cuL-QuK-hwf1VcHFy-hfGWeGoce5x8gLSEgBeRc_OPKxUf&isca=1#file28), except for whatever reason it is now highlighting in red this section of code (which is unchanged in this PR)

```
      if _, ok := neededParamsTypes[param.Name]; !ok {
                        // Ignore any missing params - this happens when extra params were
                        // passed to the task that aren't being used.
                        continue
                }
```
Screenshot 2023-03-22 at 12 08 55 PM

Screenshot 2023-03-22 at 12 09 19 PM

**Comment by EmmaMunley** [comment]
> Also not sure why the code coverage went down. I've compared the code coverage with the last PR to have file changes: #6150 and it looks like it has the exact same [code coverage](https://ffac19ab6c1d7f69330d45f61be13860ef12baf2383e8194558e81c-apidata.googleusercontent.com/download/storage/v1/b/tekton-prow/o/pr-logs%2Fpull%2Ftektoncd_pipeline%2F6150%2Fpull-tekton-pipeline-go-coverage-df%2F1633163535615791104%2Fartifacts%2Fline-cov.html?jk=AahUMlu239PxvELzbJzC_TAkhsKW2pYcTps9CQwkYDjBpJeayNOXZylnL7o4dky45nrD9dbKaDS1ElFV2wmYEnHbtehgs2XoFW6XK1FIWYpQQbIJIkyt6BiCnGz-WzDFXmViGW_HFkNptQcThofMiRThUIwKGhEcix3yCwaJtmt7Vln9QQqh2kNN8DTVXbXB6k14RkTGlagwRVKIBDV_-0Qaq4HErz-Det23TPN_sTgUFH76dZDBkjOsoe398iYmwsad4e2JLkqYMjOdbfvi6koY3troKg9Gj8gCJLu5M9PWuXyAOreMk_G7KqGk6of2DXMJ9nD9y5kUHxvczWIr4LGYoni5V1sge0BgN0GsctzRE1wItKXVUjb_REIOf54wEk5UUxNjy1H4055XsF6fXFxh5usS5HHy9SNxYOA6kgL6ZcYISQ64_mD8RH6fXqPp3gBtZKM1GE8ZpYoWeGE8l7kJXcVxpAa2--o47HUwHiY5_Ng1zyFtGMl-Tf7sHUDpY0qjWhR3PqJX13zaI06aCBPczw3aqUjM7Iic6bYVomBhpETJeUaPNds7miEMnpdR58Lza487DKdXkuzJKaIWNSqXMS3AR-i2nDrJOL8F4Ds_LhyOjDmufwfdMF096m0ANWjcM3DSgHkVDuFzzkqh8ZI4W9Jnp0SSOgiym9KAnH6LAcZsrPqyUs_2_W_rJPvLvas6r6JQ6dlLZG3Y4Oi3uSINQu6K-xHaQbxqcAPUNrH4BVpJa5sVON9xwMhB8_W7RJHxEhDv4_UUdjPVMvaOTkSF-92hKPucCb8ZbY7mP-536t7KIUmOHFGwa1JsmI4VVKfSgSIRWHJcRHcqAtOMexhKhO_oP-kQYpp6n7jdfvPj7MrPiTWgRsvZMz0LgymKtIeG2XAo7cJbZQScM72UyHXrgJpkFZbMe5yWTkt_gofz45pkScJKd5J2zI1R9Wljp70xK_5Rr1_3fXwt4dBwjeFKQAZqkzIillusJRqu2k10nC2vxa9JTZyU8JZAaOX9F0shjy8XpKpNV4GPK0sDaMP8fM8yULFcF5dSHmytclBc2YjxWUndFQNzp5zDXytYk9b3WUh8Q_jTkKFSOqlenfmRRP2QZKwhLMZYuFS51oqe1xWYb1OVcvkD--8KMmOat--N2OevKizQh2NU8bUyTMDlqlFpz1nLzpOW_nUKJ0RKej_iiF_ULix856p1aPlS0BhvXC9W6qDAGP1LRwR3Gwfkjsz0t7xdLI2BTkUMgGOCKVI5TuhP5yhNlbpFjLkbq9cuL-QuK-hwf1VcHFy-hfGWeGoce5x8gLSEgBeRc_OPKxUf&isca=1#file28), except for whatever reason it is now highlighting in red this section of code (which is unchanged in this PR)
> 
> ```
>       if _, ok := neededParamsTypes[param.Name]; !ok {
>                         // Ignore any missing params - this happens when extra params were
>                         // passed to the task that aren't being used.
>                         continue
>                 }
> ```
> 
> Screenshot 2023-03-22 at 12 08 55 PM
> 
> Screenshot 2023-03-22 at 12 09 19 PM

Added additional testing

File: pkg/reconciler/taskrun/validate_taskrun_test.go [hunk 1]
```diff
@@ -104,14 +103,17 @@ func TestValidateResolvedTask_ValidParams(t *testing.T) {
...
-	if err := ValidateResolvedTask(ctx, p, &v1beta1.Matrix{Params: m}, rtr); err != nil {
+	m := &v1beta1.Matrix{
+		Include: []v1beta1.IncludeParams{{
+			Name: "build-1",
+			Params: v1beta1.Params{{
+				Name: "include", Value: v1beta1.ParamValue{Type: v1beta1.ParamTypeString, StringVal: "string-1"},
+			}},
+		}},
+	}
+	if err := ValidateResolvedTask(ctx, p, m, rtr); err != nil {
```
Comment (jerop) [inline pkg/reconciler/taskrun/validate_taskrun_test.go:114]: this change seems to remove the test for m.Params, and add m.Include.Params -- please make sure that both are covered

File: pkg/reconciler/taskrun/validate_taskrun.go [hunk 2]
```diff
@@ -30,58 +30,41 @@ import (
 	"k8s.io/apimachinery/pkg/util/sets"
 )
 
-func validateParams(ctx context.Context, paramSpecs []v1beta1.ParamSpec, params []v1beta1.Param, matrix *v1beta1.Matrix) error {
+func validateParams(ctx context.Context, paramSpecs []v1beta1.ParamSpec, params v1beta1.Params, matrixAllParams v1beta1.Params) error {
```
Comment (lbernick) [inline pkg/reconciler/taskrun/validate_taskrun.go:33]: nit: `matrixAllParams` is a bit redundant/awkward, I'd call this `matrixParams` and add a docstring

File: pkg/reconciler/taskrun/validate_taskrun.go [hunk 3]
```diff
@@ -30,58 +30,41 @@ import (
...
 	neededParamsNames, neededParamsTypes := neededParamsNamesAndTypes(paramSpecs)
-	var matrixParams []v1beta1.Param
-	if matrix != nil {
-		matrixParams = matrix.Params
-	}
-	providedParamsNames := providedParamsNames(append(params, matrixParams...))
+	providedParams := params
+	providedParams = append(providedParams, matrixAllParams...)
```
Comment (lbernick) [inline pkg/reconciler/taskrun/validate_taskrun.go:36]: this looks like it modifies the input; can you double check that this function does not mutate its inputs?

File: pkg/reconciler/taskrun/validate_taskrun.go [hunk 4]
```diff
@@ -30,58 +30,41 @@ import (
...
 	if missingKeysObjectParamNames := MissingKeysObjectParamNames(paramSpecs, params); len(missingKeysObjectParamNames) != 0 {
 		return fmt.Errorf("missing keys for these params which are required in ParamSpec's properties %v", missingKeysObjectParamNames)
 	}
-
 	return nil
 }
-
-func neededParamsNamesAndTypes(paramSpecs []v1beta1.ParamSpec) ([]string, map[string]v1beta1.ParamType) {
-	var neededParamsNames []string
+func neededParamsNamesAndTypes(paramSpecs []v1beta1.ParamSpec) (sets.String, map[string]v1beta1.ParamType) {
```
Comment (lbernick) [inline pkg/reconciler/taskrun/validate_taskrun.go:49]: can you please add docstrings for the functions being modified?

File: pkg/reconciler/taskrun/validate_taskrun.go [hunk 5]
```diff
@@ -168,7 +151,7 @@ func findMissingKeys(neededKeys, providedKeys map[string][]string) map[string][]
 
 // ValidateResolvedTask validates task inputs, params and output matches taskrun
 func ValidateResolvedTask(ctx context.Context, params []v1beta1.Param, matrix *v1beta1.Matrix, rtr *resources.ResolvedTask) error {
-	if err := validateParams(ctx, rtr.TaskSpec.Params, params, matrix); err != nil {
+	if err := validateParams(ctx, rtr.TaskSpec.Params, params, matrix.GetAllParams()); err != nil {
```
Comment (lbernick) [inline pkg/reconciler/taskrun/validate_taskrun.go:154]: can you please update the docstring for this function?

File: pkg/reconciler/taskrun/validate_taskrun_test.go [hunk 6]
```diff
@@ -104,26 +105,37 @@ func TestValidateResolvedTask_ValidParams(t *testing.T) {
...
 		t.Fatalf("Did not expect to see error when validating TaskRun with correct params but saw %v", err)
 	}
-
 	t.Run("alpha-extra-params", func(t *testing.T) {
 		ctx := config.ToContext(ctx, &config.Config{FeatureFlags: &config.FeatureFlags{EnableAPIFields: "alpha"}})
-		extra := v1beta1.Param{
+		extraStringParam := v1beta1.Param{
```
Comment (lbernick) [inline pkg/reconciler/taskrun/validate_taskrun_test.go:125]: This test case is trying to test a lot of different things at once. Can we refactor to test one thing at a time? I think it would be good to at least move the "extra params" test into a separate test case or test class. (Also noticing we don't have coverage for "extra object param")

File: pkg/reconciler/taskrun/validate_taskrun_test.go [hunk 7]
```diff
@@ -245,6 +257,62 @@ func TestValidateResolvedTask_InvalidParams(t *testing.T) {
...
+				},
+			},
+		},
+	}
+	rtr := &resources.ResolvedTask{
+		TaskSpec: &task.Spec,
+	}
+	p := v1beta1.Params{{
+		Name:  "bar",
+		Value: *v1beta1.NewStructuredValues("somethinggood"),
```
Comment (lbernick) [inline pkg/reconciler/taskrun/validate_taskrun_test.go:285]: Is the reason this is invalid that "bar" should be an array param but is a string param? Can we use table-driven testing for these tests, with descriptive names for each test case, since they test multiple things?

**Review by lbernick — COMMENTED** [review]
thanks Emma, can you please remove the background info from the PR description, and add a bit more detail to the commit message/pr description on what validation is being added and why?

File: pkg/reconciler/taskrun/validate_taskrun.go (hunk 4 above)
Comment (EmmaMunley) [reply pkg/reconciler/taskrun/validate_taskrun.go:53]: Added docstring

File: pkg/reconciler/taskrun/validate_taskrun_test.go (hunk 7 above)
Comment (EmmaMunley) [reply pkg/reconciler/taskrun/validate_taskrun_test.go]: Refactored tests to isolate each TaskSpec

File: pkg/reconciler/taskrun/validate_taskrun.go (hunk 3 above)
Comment (EmmaMunley) [reply pkg/reconciler/taskrun/validate_taskrun.go]: I don't believe it does

File: pkg/reconciler/taskrun/validate_taskrun_test.go (hunk 6 above)
Comment (EmmaMunley) [reply pkg/reconciler/taskrun/validate_taskrun_test.go]: Created another test for testing extra params with table-driven testing. Code coverage is 100% now 🎉

File: pkg/reconciler/taskrun/validate_taskrun.go (hunk 3 above)
Comment (lbernick) [reply pkg/reconciler/taskrun/validate_taskrun.go]: my bad-- thinking of how things work in some other languages :(

File: pkg/reconciler/taskrun/validate_taskrun.go [hunk 8]
```diff
@@ -30,13 +30,13 @@ import (
 	"k8s.io/apimachinery/pkg/util/sets"
 )
 
-func validateParams(ctx context.Context, paramSpecs []v1beta1.ParamSpec, params []v1beta1.Param, matrix *v1beta1.Matrix) error {
+// validateParams validates that all Pipelink Task, Matrix.Params and Matrix.Include parameters all have values, match the specified
```
Comment (lbernick) [inline pkg/reconciler/taskrun/validate_taskrun.go:33]:
```suggestion
// validateParams validates that all Pipeline Task, Matrix.Params and Matrix.Include parameters all have values, match the specified
```

File: pkg/reconciler/taskrun/validate_taskrun.go [hunk 9]
```diff
@@ -46,42 +46,32 @@ func validateParams(ctx context.Context, paramSpecs []v1beta1.ParamSpec, params
...
-	providedParamsNames := make([]string, 0, len(params))
-	for _, param := range params {
-		providedParamsNames = append(providedParamsNames, param.Name)
-	}
-	return providedParamsNames
-}
-
-func missingParamsNames(neededParams []string, providedParams []string, paramSpecs []v1beta1.ParamSpec) []string {
-	missingParamsNames := list.DiffLeft(neededParams, providedParams)
+// missingParamsNames returns a list of missing parameter names that have not been declared with a default value
```
Comment (lbernick) [inline pkg/reconciler/taskrun/validate_taskrun.go:63]:
```suggestion
// missingParamsNames returns a slice of missing parameter names that have not been declared with a default value
```

File: pkg/reconciler/taskrun/validate_taskrun.go (hunk 5 above)
Comment (lbernick) [reply pkg/reconciler/taskrun/validate_taskrun.go:161]: I think this could use a bit more detail: what does it mean for task params and matrix params to "match" the taskrun? (I'm assuming this means all the params declared in the task are present in the taskrun and have the correct types.) Task inputs/outputs also no longer exist.

File: pkg/reconciler/taskrun/validate_taskrun_test.go [hunk 10]
```diff
@@ -104,142 +104,121 @@ func TestValidateResolvedTask_ValidParams(t *testing.T) {
...
+	}}
 	for _, tc := range tcs {
+		rtr := &resources.ResolvedTask{
+			TaskSpec: &tc.task.Spec,
+		}
 		t.Run(tc.name, func(t *testing.T) {
-			if err := ValidateResolvedTask(ctx, tc.params, tc.matrix, tc.rtr); err == nil {
-				t.Errorf("Expected to see error when validating invalid resolved TaskRun with wrong params but saw none")
+			if err := ValidateResolvedTask(ctx, tc.params, tc.matrix, rtr); err != nil {
+				t.Fatalf("Did not expect to see error when validating TaskRun with correct params but saw %v", err)
```
Comment (lbernick) [inline pkg/reconciler/taskrun/validate_taskrun_test.go:221]: Why swap to t.Fatalf?

File: pkg/reconciler/taskrun/validate_taskrun_test.go (hunk 6 above)
Comment (lbernick) [reply pkg/reconciler/taskrun/validate_taskrun_test.go]: maybe I'm missing something but I'm not seeing what happened to the tests for invalid params?

File: pkg/reconciler/taskrun/validate_taskrun_test.go [hunk 11]
```diff
@@ -104,142 +106,385 @@ func TestValidateResolvedTask_ValidParams(t *testing.T) {
...
+					Type: v1beta1.ParamTypeString,
+				}},
+			},
+		},
+		params: v1beta1.Params{{
+			Name:  "foo",
+			Value: *v1beta1.NewStructuredValues("array", "param"),
+		}},
+		matrix:  &v1beta1.Matrix{},
+		wantErr: "invalid input params for task : param types don't match the user-specified type: [foo]",
```
Comment (lbernick) [inline pkg/reconciler/taskrun/validate_taskrun_test.go:287]: Suggest including more info in this error message comparing the type received to the type specified; see https://github.com/tektoncd/pipeline/issues/6365#issuecomment-1481243995 for more detail on why.
Although this might be better in a separate PR

**Review by jerop — COMMENTED** [review]
/lgtm
